8283063: Optimize Observable{List/Set/Map}Wrapper.retainAll/removeAll

mstr2 opened the pull request:
`Observable{List/Set/Map}Wrapper.retainAll/removeAll` can be optimized for some edge cases.

1. `removeAll(c)`:
This is a no-op if 'c' is empty.
For `ObservableListWrapper`, returning early skips an object allocation. For `ObservableSetWrapper` and `ObservableMapWrapper`, returning early prevents an enumeration of the entire collection.

2. `retainAll(c)`:
This is a no-op if the backing collection is empty, or equivalent to `clear()` if `c` is empty.

I've added some tests to verify the optimized behavior for each of the three classes.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8283063](https://bugs.openjdk.org/browse/JDK-8283063): Optimize Observable{List/Set/Map}Wrapper.retainAll/removeAll


### Reviewers
 * [Nir Lisker](https://openjdk.org/census#nlisker) (@nlisker - **Reviewer**)
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/751/head:pull/751` \
`$ git checkout pull/751`

Update a local copy of the PR: \
`$ git checkout pull/751` \
`$ git pull https://git.openjdk.org/jfx.git pull/751/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 751`

View PR using the GUI difftool: \
`$ git pr show -t 751`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/751.diff">https://git.openjdk.org/jfx/pull/751.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/751#issuecomment-1065815294)

kevinrushforth commented:
/reviewers 2

yososs commented:
This is an equivalent change that partially includes the fix I proposed in #305.
It is good to apply it to Map / Set.

nlisker commented on modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableSetWrapper.java:
```diff
@@ -348,6 +348,16 @@ public boolean removeAll(Collection<?> c) {
     }
 
     private boolean removeRetain(Collection<?> c, boolean remove) {
+        if (c.isEmpty()) {
+            // Retaining an empty set is equivalent to clearing the set.
+            if (!remove && !backingSet.isEmpty()) {
+                clear();
+                return true;
+            }
+
+            return false;
+        }
```
This is good, but if `!c.isEmpty()` then we can optimize too, I think: if `backingSet.isEmpty()`, then removing/retaining all will also return `false`.

I think that doing these optimizations in the `removeAll` and `retainAll` before calling this method will be clearer, similarly to how it's done in the `List` case.

Same comment for the `Map` wrapper.

Maran23 replied:
+ 1 for this. Otherwise this changes look good to me.

nlisker reviewed (commented):
Not sure if you want to do this as part of this fix, but `ModifiableObservableListBase#addAll` can also be optimized in a similar way by checking if `c` is empty. The other methods there can also be optimized, but the `ObservableListWrapper` class overrides them anyway, so it depends on how inclusive you want this change to be.

mstr2 commented:
I've added the suggested optimization for `ModifiableObservableListBase.addAll`, as well as `ObservableSequentialListWrapper.addAll`.
Additionally, there are more tests to ensure adherence of the optimized methods to the collection contracts (NPE and IOOBE should be thrown for invalid arguments).

mstr2 commented on modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableListWrapper.java:
```diff
@@ -188,6 +192,15 @@ public boolean removeAll(Collection<?> c) {
 
     @Override
     public boolean retainAll(Collection<?> c) {
+        if (backingList.isEmpty()) {
```
Done.

mstr2 commented on modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableListWrapper.java:
```diff
@@ -169,6 +169,10 @@ public void remove(int fromIndex, int toIndex) {
 
     @Override
     public boolean removeAll(Collection<?> c) {
+        if (backingList.isEmpty() || c.isEmpty()) {
```
I've opted for implicit checks by always dereferencing `c`.

hjohn commented on modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java:
```diff
@@ -101,28 +101,46 @@ public boolean setAll(Collection<? extends E> col) {
 
     @Override
     protected void removeRange(int fromIndex, int toIndex) {
+        if (fromIndex == toIndex) {
+            if (fromIndex < 0 || fromIndex > size()) {
+                throw new IndexOutOfBoundsException("Index: " + fromIndex);
+            }
+
+            return;
+        }
```
I'm really happy you also added the index range checks, not only does it conform to the contract better, it will help expose bugs in caller code (or even the JavaFX code as Observable* classes are used in many places, and who knows what bugs might lurk in some of the more complicated classes).

I'm not quite sure how this check is suppose to work though; does this need a comment ?
```suggestion
        if (fromIndex == toIndex) {  // distinguish between the clear and remove(int) call
            if (fromIndex < 0 || fromIndex > size()) {
                throw new IndexOutOfBoundsException("Index: " + fromIndex);
            }

            return;
        }
```

I think that would be a bit too cryptic for me...

mstr2 replied:
The index check is only really necessary for the optimized code path, since calling `listIterator` later in the method would catch an invalid index. However, I've opted to always check the index at the beginning of the method.

hjohn commented on modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java:
```diff
@@ -308,43 +336,62 @@ public boolean containsAll(Collection<?> c) {
 
         @Override
         public boolean addAll(int index, Collection<? extends E> c) {
+            if (c.isEmpty()) {
+                if (index < 0 || index > sublist.size()) {
+                    throw new IndexOutOfBoundsException("Index: " + index);
+                }
+
+                return false;
+            }
+
```
Shouldn't this always check the index, not only when the collection passed in is empty?

Otherwise code like: `singleElementSubList.addAll(15, List.of("a"))` would be allowed?

```suggestion
            if (index < 0 || index > sublist.size()) {
                throw new IndexOutOfBoundsException("Index: " + index);
            }
            if (c.isEmpty()) {  // implicit null check
                return false;
            }
```

hjohn commented on modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableSequentialListWrapper.java:
```diff
@@ -175,6 +175,14 @@ public E get(int index) {
 
     @Override
     public boolean addAll(int index, Collection<? extends E> c) {
+        if (c.isEmpty()) {
+            if (index < 0 || index > size()) {
+                throw new IndexOutOfBoundsException("Index: " + index);
+            }
+
+            return false;
+        }
+
```
Shouldn't this always check the index?

```suggestion
        if (index < 0 || index > size()) {
            throw new IndexOutOfBoundsException("Index: " + index);
        }
        if (c.isEmpty()) {   // implicit null check
            return false;
        }
```

hjohn commented on modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java:
```diff
@@ -101,28 +101,46 @@ public boolean setAll(Collection<? extends E> col) {
 
     @Override
     public boolean addAll(int index, Collection<? extends E> c) {
+        if (c.isEmpty()) {
+            if (index < 0 || index > size()) {
+                throw new IndexOutOfBoundsException("Index: " + index);
+            }
+
+            return false;
+        }
+
```
Shouldn't this always check the index?

```suggestion
        if (index < 0 || index > size()) {
            throw new IndexOutOfBoundsException("Index: " + index);
        }
        if (c.isEmpty()) {  // implicit null check
            return false;
        }

```

hjohn commented on modules/javafx.base/src/test/java/test/com/sun/javafx/collections/ObservableSequentialListWrapperTest.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @Test
+    public void testAddAllWithEmptyCollectionArgumentAndInvalidIndexThrowsIOOBE() {
```
How about a variant that doesn't pass an empty collection, `-1` should still fail, and for example `100` should also fail.

mstr2 replied:
Done.

hjohn commented on modules/javafx.base/src/test/java/test/javafx/collections/ModifiableObservableListBaseTest.java:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    @Test
+    public void testAddAllWithEmptyCollectionArgumentAndInvalidIndexThrowsIOOBE() {
```
Could use variant here I think with a non empty collection, and see if it still throws IOOBE.

mstr2 replied:
Done.

hjohn commented on modules/javafx.base/src/test/java/test/javafx/collections/ModifiableObservableListBaseTest.java:
```diff
@@ -0,0 +1,249 @@
+        }
+
+        @Test
+        public void testAddAllWithEmptyCollectionArgumentAndInvalidIndexThrowsIOOBE() {
```
Could use variant here with non-empty collection, and see if it still throws IOOBE.

hjohn reviewed (changes requested):
Super happy you also added IOOBE -- collection classes are such a corner stone of Java and the observable variants fulfill that role for JavaFX -- these classes should be really conservative with the inputs they accept as any faulty inputs are likely to be bugs.

Programmers have certain expectations when they see `Set`, `List` or `Map`, and expect their inputs to be validated; when those validations are suddenly not working because the implementation happens to be of the `Observable` kind, bugs will go unnoticed.

mstr2 commented:
I've added comments in all places where implicit null checks are important to satisfy the collection contract.
In addition to that, I've also re-organized the tests using nested test classes, and added explicit null tests for all relevant code paths.

mstr2 commented on modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableMapWrapper.java:
```diff
@@ -321,20 +321,19 @@ public boolean addAll(Collection<? extends K> c) {
-        }
+            if (backingMap.isEmpty()) {
+                return false;
+            }
```
I've fixed problems like this and added comments in all places where a null check is important.

hjohn reviewed (approved):
Looks really good now!

nlisker commented on modules/javafx.base/src/test/java/test/com/sun/javafx/collections/ObservableMapWrapperTest.java:
```diff
@@ -0,0 +1,191 @@
+    class RetainAllTest {
+        @Test
+        public void testEntrySetNullArgumentThrowsNPE() {
+            var map1 = new ObservableMapWrapper<>(new HashMap<>());
+            assertThrows(NullPointerException.class, () -> map1.entrySet().retainAll((Collection<?>) null));
+
+            var map2 = new ObservableMapWrapper<>(new HashMap<>(Map.of("k0", "v0", "k1", "v1", "k2", "v2")));
+            assertThrows(NullPointerException.class, () -> map2.entrySet().retainAll((Collection<?>) null));
```
Would advise to rename `map1` to `emptyMap` and `map2` to `notEmptyMap` or the like.

Same in other places.

mstr2 replied:
Done.

nlisker reviewed (commented):
Looks good. Left a few minor comments.

mstr2 commented on modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableSequentialListWrapper.java:
```diff
@@ -175,6 +175,15 @@ public E get(int index) {
 
     @Override
     public boolean addAll(int index, Collection<? extends E> c) {
+        if (index < 0 || index > size()) {
```
In order to use `Objects.checkIndex` here, I would have to call it like this: `Objects.checkIndex(index, size() + 1)`.
`size() + 1` is necessary because it is valid to call this method with an `index` equal to the size of the list.

However, the exception message that would be produced for an out-of-bounds index would be incorrect, for example when adding elements at index 20 to a list that already contains 10 elements: "Index 20 out of bounds for length 11". Obviously, a list of 10 elements doesn't have a length of 11.

mstr2 commented on modules/javafx.base/src/test/java/test/com/sun/javafx/collections/ObservableListWrapperTest.java:
```diff
@@ -0,0 +1,95 @@
+                }
+            });
+
+            list.removeAll(Collections.<String>emptyList());
```
Done.

mstr2 commented on modules/javafx.base/src/test/java/test/com/sun/javafx/collections/ObservableSequentialListWrapperTest.java:
```diff
@@ -0,0 +1,72 @@
+            }
+        };
+
+        assertDoesNotThrow(() -> list.addAll(Collections.emptyList()));
```
Good catch, I've removed the method test.

mstr2 commented on modules/javafx.base/src/test/java/test/com/sun/javafx/collections/ObservableMapWrapperTest.java:
```diff
@@ -0,0 +1,191 @@
+        }
+
+        @Test
+        public void testValueSetNullArgumentThrowsNPE() {
```
Fixed that.

mstr2 commented on modules/javafx.base/src/test/java/test/com/sun/javafx/collections/ObservableListWrapperTest.java:
```diff
@@ -0,0 +1,95 @@
+public class ObservableListWrapperTest {
+
+    @Nested
+    class RemoveAllTest {
```
I see that all over the JavaFX codebase. Do we have any guidance for that?

kevinrushforth replied:
We're inconsistent on this. Since this is a new class, I'd probably lean towards adding the blank line.

mstr2 commented:
@hjohn Can you re-approve?
